File: keys-personal-assist-ui/src/types/liability.ts
import type { CompoundingFrequency } from './asset';

export interface LiabilitySubcategory {
  id: string;
  categoryId: string;
  name: string;
  code: string;
  description: string | null;
  valuationType: 'VALUE_BASED' | 'UNIT_BASED';
  hasInterest: boolean;
  hasMaturity: boolean;
}

export interface LiabilityCategory {
  id: string;
  name: string;
  code: 'SECURED_LOAN' | 'UNSECURED_LOAN' | 'REVOLVING_CREDIT' | 'OTHER';
  description: string | null;
  subcategories: LiabilitySubcategory[];
}

export interface LiabilityInterestDetails {
  interestRate: number;
  compounding: CompoundingFrequency;
  emiAmount?: number | null;
  maturityDate?: string | null;
}

export interface Liability {
  id: string;
  categoryId: string;
  categoryName: string;
  categoryCode: 'SECURED_LOAN' | 'UNSECURED_LOAN' | 'REVOLVING_CREDIT' | 'OTHER';
  name: string;
  subcategoryId: string | null;
  originalValue: number;
  currentValue: number; // outstanding balance
  interestRate: number | null;
  interestCompounding: CompoundingFrequency | null;
  emiAmount: number | null;
  maturityDate: string | null;
  notes: string | null;
  totalRepaid: number;
  accumulated_interest?: number; // wait, let's match camelCase from API or snake_case
  // Let's check schemas/liability.py: accumulatedInterest is camelCased due to BaseSchema's alias_generator=to_camel
  accumulatedInterest: number;
  progressPct: number;
  createdAt: string; // ISO datetime
  updatedAt: string; // ISO datetime
}

export interface LiabilityRequest {
  categoryId: string;
  name: string;
  subcategoryId?: string | null;
  initialAmount: number;
  initialDate?: string | null;
  interestDetails?: LiabilityInterestDetails | null;
  notes?: string | null;
}

export interface LiabilityUpdateRequest {
  categoryId?: string | null;
  name?: string | null;
  subcategoryId?: string | null;
  interestDetails?: LiabilityInterestDetails | null;
  notes?: string | null;
}

export interface LiabilityTransaction {
  id: string;
  liabilityId: string;
  transactionType: 'BORROW' | 'REPAY' | 'REVALUE';
  amount: number;
  transactionDate: string; // ISO datetime
  description: string | null;
}

export interface LiabilityTransactionRequest {
  transactionType: 'BORROW' | 'REPAY' | 'REVALUE';
  amount: number;
  transactionDate?: string | null; // ISO datetime
  description?: string | null;
}

export interface LiabilityCategorySummary {
  categoryId: string;
  categoryName: string;
  categoryCode: 'SECURED_LOAN' | 'UNSECURED_LOAN' | 'REVOLVING_CREDIT' | 'OTHER';
  totalOriginal: number;
  totalOutstanding: number;
  totalRepaid: number;
  accumulatedInterest: number;
}

export interface LiabilitySummary {
  totalOriginal: number;
  totalOutstanding: number;
  totalRepaid: number;
  accumulatedInterest: number;
  categoryBreakdowns: LiabilityCategorySummary[];
}

export interface LiabilityProjectionPoint {
  date: string;
  idealBalance: number;
  actualBalance: number | null;
  idealInterestPaid: number;
  actualInterestPaid: number | null;
}

export interface LiabilityProjectionMetrics {
  idealTenureMonths: number;
  remainingTenureMonths: number;
  tenureSavedMonths: number;
  totalInterestIdeal: number;
  totalInterestProjected: number;
  interestSaved: number;
  idealEndDate: string | null;
  projectedEndDate: string | null;
}

export interface LiabilityProjections {
  metrics: LiabilityProjectionMetrics;
  projectionPoints: LiabilityProjectionPoint[];
}

File: services/expense-manager-service/app/routers/v1/schemas/liability.py
"""Schemas for liability API endpoints."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from app.entities.models.liability import CompoundingFrequency, LiabilityTransactionType


class BaseSchema(BaseModel):
    """Base schema with camelCase settings."""

    model_config = ConfigDict(
        populate_by_name=True,
        alias_generator=to_camel,
    )


# Sub-schemas


class LiabilityInterestDetailsSchema(BaseSchema):
    """Schema for interest-bearing liability fields."""

    interest_rate: float = Field(gt=0, description="Annual interest rate (%)")
    compounding: CompoundingFrequency = Field(description="Compounding frequency")
    emi_amount: float | None = Field(default=None, gt=0, description="Scheduled monthly EMI amount (INR)")
    emi_start_date: datetime | None = Field(default=None, description="Date when EMI repayments officially begin")
    maturity_date: datetime | None = Field(default=None, description="Maturity date of the liability")


# Request schemas


class LiabilityRequest(BaseSchema):
    """Schema for adding a new liability."""

    category_id: str = Field(description="ID of the parent category")
    name: str = Field(description="Name of the liability")
    subcategory_id: str | None = Field(default=None, description="ID of the subcategory")
    initial_amount: float = Field(gt=0, description="Amount for the initial BORROW transaction (INR)")
    initial_date: datetime | None = Field(default=None, description="Timestamp of initial borrowing")
    interest_details: LiabilityInterestDetailsSchema | None = Field(
        default=None, description="Interest fields. Required for interest-bearing subcategories."
    )
    notes: str | None = Field(default=None, description="Additional notes/remarks")


class LiabilityUpdateRequest(BaseSchema):
    """Schema for partially updating liability metadata (PATCH)."""

    category_id: str | None = Field(default=None, description="ID of the parent category")
    name: str | None = Field(default=None, description="Name of the liability")
    subcategory_id: str | None = Field(default=None, description="ID of the subcategory")
    interest_details: LiabilityInterestDetailsSchema | None = Field(
        default=None, description="Updated interest fields. Pass null to clear interest tracking."
    )
    notes: str | None = Field(default=None, description="Additional notes/remarks")


class LiabilityTransactionRequest(BaseSchema):
    """Schema for logging a new transaction."""

    transaction_type: LiabilityTransactionType = Field(description="Transaction type: BORROW, REPAY, or REVALUE")
    amount: float = Field(gt=0, description="Total INR amount of the transaction")
    transaction_date: datetime | None = Field(default=None, description="Transaction timestamp")
    description: str | None = Field(default=None, description="Audit remark")


# Response schemas


class LiabilitySubcategoryResponse(BaseSchema):
    """Response schema for liability subcategory."""

    id: str
    category_id: str
    name: str
    code: str
    description: str | None = Field(default=None)
    valuation_type: str
    has_interest: bool
    has_maturity: bool


class LiabilityCategoryResponse(BaseSchema):
    """Response schema for liability category."""

    id: str
    name: str
    code: str
    description: str | None
    subcategories: list[LiabilitySubcategoryResponse] = Field(default_factory=list)


class LiabilityResponse(BaseSchema):
    """Response schema for liability details."""

    id: str
    category_id: str
    category_name: str
    category_code: str
    name: str
    subcategory_id: str | None
    original_value: float
    current_value: float
    interest_rate: float | None
    interest_compounding: CompoundingFrequency | None
    emi_amount: float | None
    emi_start_date: datetime | None
    maturity_date: datetime | None
    notes: str | None
    total_repaid: float
    accumulated_interest: float
    progress_pct: float
    remaining_tenure_months: int | None = Field(default=None, description="Estimated remaining tenure in months")
    created_at: datetime
    updated_at: datetime


class LiabilityTransactionResponse(BaseSchema):
    """Response schema for liability transaction."""

    id: str
    liability_id: str
    transaction_type: LiabilityTransactionType
    amount: float
    transaction_date: datetime
    description: str | None


class LiabilityCategorySummaryResponse(BaseSchema):
    """Response schema for category aggregate summary."""

    category_id: str
    category_name: str
    category_code: str
    total_original: float
    total_outstanding: float
    total_repaid: float
    accumulated_interest: float


class LiabilitySummaryResponse(BaseSchema):
    """Response schema for full liabilities dashboard overview."""

    total_original: float
    total_outstanding: float
    total_repaid: float
    accumulated_interest: float
    category_breakdowns: list[LiabilityCategorySummaryResponse]


class LiabilityProjectionPointResponse(BaseSchema):
    """API response schema for a single projection monthly data point."""

    date: str
    ideal_balance: float
    actual_balance: float | None = None
    ideal_interest_paid: float
    actual_interest_paid: float | None = None


class LiabilityProjectionMetricsResponse(BaseSchema):
    """API response schema for liability payoff projection metrics."""

    ideal_tenure_months: int
    remaining_tenure_months: int | None = None
    tenure_saved_months: int | None = None
    total_interest_ideal: float
    total_interest_projected: float
    interest_saved: float
    ideal_end_date: datetime | None = None
    projected_end_date: datetime | None = None


class LiabilityProjectionsResponse(BaseSchema):
    """API response schema for complete liability projections."""

    metrics: LiabilityProjectionMetricsResponse
    projection_points: list[LiabilityProjectionPointResponse]
